File: src/chroma/code/python/error-handling.py
import json
import os
import time

import chromadb
from chromadb.errors import NotFoundError

# @snippet:connection-errors


def connect_with_retry(max_retries: int = 3) -> chromadb.ClientAPI:
    """Connect to Chroma with exponential backoff retry."""
    client = chromadb.Client()

    for attempt in range(1, max_retries + 1):
        try:
            # heartbeat() verifies the connection is working
            client.heartbeat()
            return client
        except Exception as e:
            if attempt == max_retries:
                raise ConnectionError(
                    f"Failed to connect to Chroma after {max_retries} attempts: {e}"
                )
            # Exponential backoff: 1s, 2s, 4s...
            delay = 2 ** (attempt - 1)
            time.sleep(delay)

    raise ConnectionError("Unreachable")


# @end

# @snippet:collection-not-found

client = chromadb.Client()

# get_collection raises NotFoundError if the collection doesn't exist
try:
    collection = client.get_collection(name="my_collection")
except NotFoundError:
    print("Collection not found, creating it...")
    collection = client.create_collection(name="my_collection")

# @end

# @snippet:safe-collection-access

client = chromadb.Client()

# Preferred: get_or_create_collection never raises NotFoundError
collection = client.get_or_create_collection(name="my_collection")

# Check if results exist before accessing
results = collection.query(
    query_texts=["search query"],
    n_results=5,
)

if results["documents"] and len(results["documents"][0]) > 0:
    first_doc = results["documents"][0][0]
    # Safe to use first_doc
else:
    # No results found
    pass

# @end

# @snippet:validation-errors

client = chromadb.Client()


def validate_document(doc: str) -> bool:
    """Validate document size (16KB limit, recommend < 8KB)."""
    byte_size = len(doc.encode("utf-8"))
    return byte_size <= 16384


def validate_metadata(metadata: dict) -> bool:
    """Validate metadata size (4KB limit, 32 keys max)."""
    if len(metadata.keys()) > 32:
        return False

    json_size = len(json.dumps(metadata).encode("utf-8"))
    return json_size <= 4096


def safe_add(
    collection_name: str,
    ids: list[str],
    documents: list[str],
    metadatas: list[dict] | None = None,
):
    """Add documents with pre-validation."""
    # Pre-validate documents
    for doc in documents:
        if not validate_document(doc):
            raise ValueError("Document exceeds 16KB limit")

    # Pre-validate metadata
    if metadatas:
        for meta in metadatas:
            if not validate_metadata(meta):
                raise ValueError("Metadata exceeds limits (4KB or 32 keys)")

    collection = client.get_or_create_collection(name=collection_name)

    try:
        collection.add(ids=ids, documents=documents, metadatas=metadatas)
    except ValueError as e:
        # Handle specific validation errors
        error_msg = str(e).lower()
        if "dimension" in error_msg:
            raise ValueError("Embedding dimensions do not match collection") from e
        if "duplicate" in error_msg:
            raise ValueError("Duplicate IDs in batch") from e
        raise


# @end

# @snippet:batch-operations

client = chromadb.Client()


def batch_add(
    collection_name: str,
    ids: list[str],
    documents: list[str],
    batch_size: int = 100,
) -> dict:
    """Process documents in batches to handle partial failures."""
    collection = client.get_or_create_collection(name=collection_name)

    failures = []

    for i in range(0, len(ids), batch_size):
        batch_ids = ids[i : i + batch_size]
        batch_docs = documents[i : i + batch_size]

        try:
            collection.add(ids=batch_ids, documents=batch_docs)
        except Exception as e:
            # Log failure but continue with other batches
            failures.append({"index": i, "error": str(e)})

    if failures:
        print(f"{len(failures)} batches failed: {failures}")

    total_batches = (len(ids) + batch_size - 1) // batch_size
    return {"total_batches": total_batches, "failures": failures}


# @end

# @snippet:cloud-errors


def get_cloud_config() -> dict:
    """Verify environment variables before creating client."""
    api_key = os.environ.get("CHROMA_API_KEY")
    tenant = os.environ.get("CHROMA_TENANT")
    database = os.environ.get("CHROMA_DATABASE")

    missing = []
    if not api_key:
        missing.append("CHROMA_API_KEY")
    if not tenant:
        missing.append("CHROMA_TENANT")
    if not database:
        missing.append("CHROMA_DATABASE")

    if missing:
        raise EnvironmentError(
            f"Missing required environment variables: {', '.join(missing)}"
        )

    return {"api_key": api_key, "tenant": tenant, "database": database}


def create_cloud_client() -> chromadb.ClientAPI:
    """Create CloudClient with error handling."""
    config = get_cloud_config()

    client = chromadb.CloudClient(
        api_key=config["api_key"],
        tenant=config["tenant"],
        database=config["database"],
    )

    try:
        # Verify connection and authentication
        client.heartbeat()
        return client
    except Exception as e:
        error_msg = str(e).lower()
        if "401" in error_msg or "unauthorized" in error_msg:
            raise PermissionError("Invalid or expired API key") from e
        if "404" in error_msg or "not found" in error_msg:
            raise NotFoundError(
                "Tenant or database not found - check configuration"
            ) from e
        if "429" in error_msg or "rate" in error_msg:
            raise RuntimeError("Rate limit exceeded - implement backoff") from e
        raise


# @end

File: src/chroma/code/typescript/error-handling.ts
import { ChromaClient, CloudClient } from 'chromadb';
import { DefaultEmbeddingFunction } from '@chroma-core/default-embed';

// @snippet:connection-errors

async function connectWithRetry(maxRetries = 3): Promise<ChromaClient> {
  const client = new ChromaClient();

  for (let attempt = 1; attempt <= maxRetries; attempt++) {
    try {
      // heartbeat() verifies the connection is working
      await client.heartbeat();
      return client;
    } catch (error) {
      if (attempt === maxRetries) {
        throw new Error(
          `Failed to connect to Chroma after ${maxRetries} attempts: ${error}`
        );
      }
      // Exponential backoff: 1s, 2s, 4s...
      const delay = Math.pow(2, attempt - 1) * 1000;
      await new Promise((resolve) => setTimeout(resolve, delay));
    }
  }
  throw new Error('Unreachable');
}

// @end

// @snippet:collection-not-found

const client = new ChromaClient();
const embeddingFunction = new DefaultEmbeddingFunction();

// getCollection throws if the collection doesn't exist
try {
  const collection = await client.getCollection({
    name: 'my_collection',
    embeddingFunction,
  });
} catch (error) {
  if (error instanceof Error && error.message.includes('does not exist')) {
    console.log('Collection not found, creating it...');
    // Handle missing collection
  } else {
    throw error; // Re-throw unexpected errors
  }
}

// @end

// @snippet:safe-collection-access

const client2 = new ChromaClient();
const embeddingFunction2 = new DefaultEmbeddingFunction();

// Preferred: getOrCreateCollection never throws "not found"
const collection = await client2.getOrCreateCollection({
  name: 'my_collection',
  embeddingFunction: embeddingFunction2,
});

// Check if results exist before accessing
const results = await collection.query({
  queryTexts: ['search query'],
  nResults: 5,
});

if (results.documents[0] && results.documents[0].length > 0) {
  const firstDoc = results.documents[0][0];
  // Safe to use firstDoc
} else {
  // No results found
}

// @end

// @snippet:validation-errors

const client3 = new ChromaClient();
const embeddingFunction3 = new DefaultEmbeddingFunction();

// Validate document size before adding (16KB limit, recommend < 8KB)
function validateDocument(doc: string): boolean {
  const byteSize = new TextEncoder().encode(doc).length;
  return byteSize <= 16384;
}

// Validate metadata size (4KB limit, 32 keys max)
function validateMetadata(
  metadata: Record<string, string | number | boolean>
): boolean {
  const keys = Object.keys(metadata);
  if (keys.length > 32) return false;

  const jsonSize = new TextEncoder().encode(JSON.stringify(metadata)).length;
  return jsonSize <= 4096;
}

async function safeAdd(
  collectionName: string,
  ids: string[],
  documents: string[],
  metadatas?: Record<string, string | number | boolean>[]
) {
  // Pre-validate
  for (const doc of documents) {
    if (!validateDocument(doc)) {
      throw new Error(`Document exceeds 16KB limit`);
    }
  }

  if (metadatas) {
    for (const meta of metadatas) {
      if (!validateMetadata(meta)) {
        throw new Error(`Metadata exceeds limits (4KB or 32 keys)`);
      }
    }
  }

  const collection = await client3.getOrCreateCollection({
    name: collectionName,
    embeddingFunction: embeddingFunction3,
  });

  try {
    await collection.add({ ids, documents, metadatas });
  } catch (error) {
    // Handle specific validation errors from server
    if (error instanceof Error) {
      if (error.message.includes('dimension')) {
        throw new Error('Embedding dimensions do not match collection');
      }
      if (error.message.includes('duplicate')) {
        throw new Error('Duplicate IDs in batch');
      }
    }
    throw error;
  }
}

// @end

// @snippet:batch-operations

const client4 = new ChromaClient();
const embeddingFunction4 = new DefaultEmbeddingFunction();

// Process documents in batches to avoid memory issues and partial failures
async function batchAdd(
  collectionName: string,
  ids: string[],
  documents: string[],
  batchSize = 100
) {
  const collection = await client4.getOrCreateCollection({
    name: collectionName,
    embeddingFunction: embeddingFunction4,
  });

  const failures: { index: number; error: string }[] = [];

  for (let i = 0; i < ids.length; i += batchSize) {
    const batchIds = ids.slice(i, i + batchSize);
    const batchDocs = documents.slice(i, i + batchSize);

    try {
      await collection.add({
        ids: batchIds,
        documents: batchDocs,
      });
    } catch (error) {
      // Log failure but continue with other batches
      failures.push({
        index: i,
        error: error instanceof Error ? error.message : String(error),
      });
    }
  }

  if (failures.length > 0) {
    console.error(`${failures.length} batches failed:`, failures);
  }

  return { totalBatches: Math.ceil(ids.length / batchSize), failures };
}

// @end

// @snippet:cloud-errors

// Verify environment variables before creating client
function getCloudConfig() {
  const apiKey = process.env.CHROMA_API_KEY;
  const tenant = process.env.CHROMA_TENANT;
  const database = process.env.CHROMA_DATABASE;

  const missing: string[] = [];
  if (!apiKey) missing.push('CHROMA_API_KEY');
  if (!tenant) missing.push('CHROMA_TENANT');
  if (!database) missing.push('CHROMA_DATABASE');

  if (missing.length > 0) {
    throw new Error(
      `Missing required environment variables: ${missing.join(', ')}`
    );
  }

  return { apiKey, tenant, database };
}

async function createCloudClient() {
  const config = getCloudConfig();

  const client = new CloudClient({
    apiKey: config.apiKey,
    tenant: config.tenant,
    database: config.database,
  });

  try {
    // Verify connection and authentication
    await client.heartbeat();
    return client;
  } catch (error) {
    if (error instanceof Error) {
      if (
        error.message.includes('401') ||
        error.message.includes('Unauthorized')
      ) {
        throw new Error('Invalid or expired API key');
      }
      if (
        error.message.includes('404') ||
        error.message.includes('not found')
      ) {
        throw new Error('Tenant or database not found - check configuration');
      }
      if (error.message.includes('429') || error.message.includes('rate')) {
        throw new Error('Rate limit exceeded - implement backoff');
      }
    }
    throw error;
  }
}

// @end
